File: pkg/sextantproto/envelope.go
package sextantproto

import (
	"encoding/json"
	"fmt"
	"time"

	"github.com/google/uuid"
)

// Envelope is the universal wrapper for every message on the sextant bus.
// Every field listed in specs/protocols/envelope-schema.md is present here.
// JSON tags match the wire format exactly.
//
// Construct via NewEnvelope so trace IDs are populated; the zero value is
// invalid because TraceID and SpanID are required (envelope-schema.md §2).
type Envelope struct {
	ID           uuid.UUID `json:"id"`
	Ts           Timestamp `json:"ts"`
	ProtoVersion string    `json:"proto_version"`

	From Address  `json:"from"`
	To   *Address `json:"to,omitempty"`

	TraceID      uuid.UUID  `json:"trace_id"`
	SpanID       uuid.UUID  `json:"span_id"`
	ParentSpanID *uuid.UUID `json:"parent_span_id,omitempty"`

	IdempotencyKey *string `json:"idempotency_key,omitempty"`
	ReplyTo        *string `json:"reply_to,omitempty"`

	Kind    Kind            `json:"kind"`
	Payload json.RawMessage `json:"payload"`
}

// Address identifies an envelope's sender or recipient.
//
// Kind is a closed enum: "agent" | "operator" | "daemon" | "ui" | "external".
// ID is the UUID for kinds that have one; "daemon-<host>" for daemons.
type Address struct {
	Kind AddressKind `json:"kind"`
	ID   string      `json:"id"`
	Host *string     `json:"host,omitempty"`
}

// AddressKind enumerates the legal Address.Kind values.
type AddressKind string

const (
	AddressAgent    AddressKind = "agent"
	AddressOperator AddressKind = "operator"
	AddressDaemon   AddressKind = "daemon"
	AddressUI       AddressKind = "ui"
	AddressExternal AddressKind = "external"
)

// IsValid reports whether k is a recognized AddressKind.
func (k AddressKind) IsValid() bool {
	switch k {
	case AddressAgent, AddressOperator, AddressDaemon, AddressUI, AddressExternal:
		return true
	default:
		return false
	}
}

// AllAddressKinds returns every AddressKind in canonical order. Used by
// code generation (the wire.json manifest + generated TS constants) and
// tests; do not rely on it for hot paths.
func AllAddressKinds() []AddressKind {
	return []AddressKind{
		AddressAgent,
		AddressOperator,
		AddressDaemon,
		AddressUI,
		AddressExternal,
	}
}

// Kind is the envelope discriminator. Every value here corresponds to a
// payload type in this package.
type Kind string

const (
	KindAgentFrame        Kind = "agent_frame"
	KindLifecycle         Kind = "lifecycle"
	KindAudit             Kind = "audit"
	KindTelemetrySpan     Kind = "telemetry_span"
	KindTelemetryMetric   Kind = "telemetry_metric"
	KindTelemetryLog      Kind = "telemetry_log"
	KindUserInputRequest  Kind = "user_input_request"
	KindUserInputResponse Kind = "user_input_response"
	KindRPCRequest        Kind = "rpc_request"
	KindRPCResponse       Kind = "rpc_response"
	KindHeartbeat         Kind = "heartbeat"
)

// AllKinds returns every envelope kind in canonical order. Used by code
// generation and tests; do not rely on it for hot paths.
func AllKinds() []Kind {
	return []Kind{
		KindAgentFrame,
		KindLifecycle,
		KindAudit,
		KindTelemetrySpan,
		KindTelemetryMetric,
		KindTelemetryLog,
		KindUserInputRequest,
		KindUserInputResponse,
		KindRPCRequest,
		KindRPCResponse,
		KindHeartbeat,
	}
}

// NewEnvelope returns a root envelope populated with a fresh ID,
// current timestamp, the current ProtoVersion, and trace fields set up
// for a root span (TraceID = ID, SpanID = fresh UUID, no parent).
//
// Payload must already be JSON-marshalled; use NewEnvelopeWith for a
// helper that marshals the payload for you.
func NewEnvelope(kind Kind, from Address, payload json.RawMessage) Envelope {
	id := uuid.New()
	return Envelope{
		ID:           id,
		Ts:           NowTimestamp(),
		ProtoVersion: ProtoVersion,
		From:         from,
		TraceID:      id,
		SpanID:       uuid.New(),
		Kind:         kind,
		Payload:      payload,
	}
}

// NewEnvelopeWith marshals payload to JSON and wraps it in a root envelope.
// Returns an error if marshalling fails.
func NewEnvelopeWith(kind Kind, from Address, payload any) (Envelope, error) {
	raw, err := json.Marshal(payload)
	if err != nil {
		return Envelope{}, fmt.Errorf("marshal envelope payload for kind %q: %w", kind, err)
	}
	return NewEnvelope(kind, from, raw), nil
}

// Child returns a derived envelope that shares the parent's TraceID and
// references the parent's SpanID as its parent. ID, SpanID, and Ts are
// fresh. Kind, From, and Payload are caller-supplied.
func (e Envelope) Child(kind Kind, from Address, payload json.RawMessage) Envelope {
	parent := e.SpanID
	return Envelope{
		ID:           uuid.New(),
		Ts:           NowTimestamp(),
		ProtoVersion: ProtoVersion,
		From:         from,
		TraceID:      e.TraceID,
		SpanID:       uuid.New(),
		ParentSpanID: &parent,
		Kind:         kind,
		Payload:      payload,
	}
}

// Validate checks the structural invariants required by envelope-schema.md:
// non-nil ID, TraceID, and SpanID; recognized Kind and From.Kind;
// ProtoVersion present.
func (e Envelope) Validate() error {
	if e.ID == uuid.Nil {
		return fmt.Errorf("envelope: id is nil")
	}
	if e.TraceID == uuid.Nil {
		return fmt.Errorf("envelope: trace_id is nil (required on every envelope)")
	}
	if e.SpanID == uuid.Nil {
		return fmt.Errorf("envelope: span_id is nil (required on every envelope)")
	}
	if e.ProtoVersion == "" {
		return fmt.Errorf("envelope: proto_version is empty")
	}
	if e.Kind == "" {
		return fmt.Errorf("envelope: kind is empty")
	}
	if !e.From.Kind.IsValid() {
		return fmt.Errorf("envelope: from.kind %q is not a recognized AddressKind", e.From.Kind)
	}
	if e.Ts.IsZero() {
		return fmt.Errorf("envelope: ts is zero")
	}
	return nil
}

// Timestamp is a wall-clock instant with microsecond precision, encoded
// as RFC 3339 with a 6-digit fractional second component. Wire format is
// stable across languages; ClickHouse columns use DateTime64(6).
type Timestamp struct {
	time.Time
}

// NowTimestamp returns the current time truncated to microsecond precision.
func NowTimestamp() Timestamp {
	return Timestamp{time.Now().UTC().Truncate(time.Microsecond)}
}

// AtTimestamp wraps t and truncates to microsecond precision.
func AtTimestamp(t time.Time) Timestamp {
	return Timestamp{t.UTC().Truncate(time.Microsecond)}
}

// MarshalJSON emits the timestamp as an RFC 3339 string truncated to
// microseconds.
func (t Timestamp) MarshalJSON() ([]byte, error) {
	if t.IsZero() {
		return []byte(`null`), nil
	}
	return []byte(`"` + t.Time.UTC().Format("2006-01-02T15:04:05.000000Z07:00") + `"`), nil
}

// UnmarshalJSON accepts RFC 3339 strings with up to nanosecond precision
// and JSON null. The parsed time is truncated to microseconds so a
// round-trip is stable.
func (t *Timestamp) UnmarshalJSON(b []byte) error {
	if string(b) == "null" || string(b) == `""` {
		*t = Timestamp{}
		return nil
	}
	var s string
	if err := json.Unmarshal(b, &s); err != nil {
		return fmt.Errorf("timestamp: not a string: %w", err)
	}
	parsed, err := time.Parse(time.RFC3339Nano, s)
	if err != nil {
		return fmt.Errorf("timestamp: parse %q: %w", s, err)
	}
	*t = Timestamp{parsed.UTC().Truncate(time.Microsecond)}
	return nil
}

File: clients/typescript/src/envelope.ts
/**
 * Envelope construction + validation helpers.
 *
 * Mirrors pkg/sextantproto/envelope.go: NewEnvelope (root span),
 * Child (causally-linked span), Validate (structural invariants).
 * Wire format is JSON; see specs/protocols/envelope-schema.md.
 */

import { randomUUID } from "node:crypto";

import type { Address, Envelope } from "./types.generated.js";

/** Envelope protocol version emitted by this client (matches Go ProtoVersion). */
export const PROTO_VERSION = "0.5.0";

/** Recognized envelope kinds. Mirrors sextantproto.Kind. */
export const KIND_AGENT_FRAME = "agent_frame";
export const KIND_LIFECYCLE = "lifecycle";
export const KIND_AUDIT = "audit";
export const KIND_TELEMETRY_SPAN = "telemetry_span";
export const KIND_TELEMETRY_METRIC = "telemetry_metric";
export const KIND_TELEMETRY_LOG = "telemetry_log";
export const KIND_USER_INPUT_REQUEST = "user_input_request";
export const KIND_USER_INPUT_RESPONSE = "user_input_response";
export const KIND_RPC_REQUEST = "rpc_request";
export const KIND_RPC_RESPONSE = "rpc_response";
export const KIND_HEARTBEAT = "heartbeat";

/** Recognized Address kinds. Mirrors sextantproto.AddressKind. */
export const ADDRESS_AGENT = "agent";
export const ADDRESS_OPERATOR = "operator";
export const ADDRESS_DAEMON = "daemon";
export const ADDRESS_UI = "ui";
export const ADDRESS_EXTERNAL = "external";

const VALID_ADDRESS_KINDS = new Set([
  ADDRESS_AGENT,
  ADDRESS_OPERATOR,
  ADDRESS_DAEMON,
  ADDRESS_UI,
  ADDRESS_EXTERNAL,
]);

/**
 * Format a Date (or "now") as the wire timestamp form: RFC 3339 with
 * exactly 6 fractional digits and a `Z` suffix. Mirrors
 * sextantproto.Timestamp.MarshalJSON.
 */
export function formatTimestamp(d: Date = new Date()): string {
  const iso = d.toISOString(); // e.g. "2026-05-24T12:34:56.789Z"
  // toISOString gives 3-digit fractional seconds. Pad to 6 to match Go.
  const dot = iso.indexOf(".");
  if (dot < 0) {
    return iso.replace("Z", ".000000Z");
  }
  const baseEnd = iso.indexOf("Z");
  const frac = iso.slice(dot + 1, baseEnd);
  const padded = (frac + "000000").slice(0, 6);
  return `${iso.slice(0, dot)}.${padded}Z`;
}

/** Parse a wire timestamp into a JS Date (microsecond precision lost — JS is ms). */
export function parseTimestamp(s: string): Date {
  return new Date(s);
}

/**
 * Build a root envelope (no causal parent). TraceID is set to the
 * envelope ID, SpanID is fresh — matches pkg/sextantproto.NewEnvelope.
 */
export function newEnvelope(
  kind: string,
  from: Address,
  payload: unknown,
): Envelope {
  const id = randomUUID();
  return {
    id,
    ts: formatTimestamp(),
    proto_version: PROTO_VERSION,
    from,
    trace_id: id,
    span_id: randomUUID(),
    kind,
    payload,
  };
}

/**
 * Build a child envelope under `parent`, preserving trace_id and
 * referencing parent.span_id as parent_span_id. ID, span_id, ts are
 * fresh. Mirrors `(Envelope).Child` in Go.
 */
export function childEnvelope(
  parent: Envelope,
  kind: string,
  from: Address,
  payload: unknown,
): Envelope {
  return {
    id: randomUUID(),
    ts: formatTimestamp(),
    proto_version: PROTO_VERSION,
    from,
    trace_id: parent.trace_id,
    span_id: randomUUID(),
    parent_span_id: parent.span_id,
    kind,
    payload,
  };
}

/**
 * Throw if envelope violates the structural invariants required by
 * specs/protocols/envelope-schema.md. Mirrors `(Envelope).Validate`.
 *
 * Called by `publish` so a malformed envelope fails on the publisher,
 * not on every downstream consumer.
 */
export function validateEnvelope(env: Envelope): void {
  if (!env.id) throw new Error("envelope: id is empty");
  if (!env.trace_id) throw new Error("envelope: trace_id is empty (required on every envelope)");
  if (!env.span_id) throw new Error("envelope: span_id is empty (required on every envelope)");
  if (!env.proto_version) throw new Error("envelope: proto_version is empty");
  if (!env.kind) throw new Error("envelope: kind is empty");
  if (!env.from || !VALID_ADDRESS_KINDS.has(env.from.kind)) {
    throw new Error(`envelope: from.kind ${JSON.stringify(env.from?.kind)} is not a recognized AddressKind`);
  }
  if (!env.ts) throw new Error("envelope: ts is empty");
}

/**
 * Parse and validate a JSON-encoded envelope from a NATS message
 * payload. Returns the envelope on success; throws on decode or
 * validation failure.
 *
 * Subscribers use this; on failure the message is Term'd at the
 * JetStream level so a malformed event is reported once.
 */
export function decodeEnvelope(data: Uint8Array): Envelope {
  const text = new TextDecoder().decode(data);
  let env: Envelope;
  try {
    env = JSON.parse(text) as Envelope;
  } catch (err) {
    throw new Error(
      `unmarshal envelope: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
  validateEnvelope(env);
  return env;
}

/**
 * Encode an envelope to a UTF-8 byte array suitable for nc.publish.
 *
 * If `env.ts` is empty, it's set to `formatTimestamp()` first. Same
 * for `proto_version`. Matches Publish's small-fix behaviour in Go.
 */
export function encodeEnvelope(env: Envelope): Uint8Array {
  if (!env.ts) env.ts = formatTimestamp();
  if (!env.proto_version) env.proto_version = PROTO_VERSION;
  return new TextEncoder().encode(JSON.stringify(env));
}
